src/backend/cuda/kernel/meanshift.hpp

```diff
@@ -12,200 +12,152 @@
-    float tmpclrs[channels];
+    const int dim0LenLmt = in.dims[0]-1;
+    const int dim1LenLmt = in.dims[1]-1;
 
-    // clear means and centers for this pixel
-#pragma unroll
-    for(int ch=0; ch<channels; ++ch) {
-        means[ch] = 0.0f;
-        centers[ch] = shrdMem[lIdx(i, j, shrdLen, 1)+ch*schStride];
-    }
+    // scope of meanshift iterations begin
+    for (uint it=0; it<iter; ++it) {
 
-    // scope of meanshift iterationd begin
-    for(uint it=0; it<iter; ++it) {
+        int ocj = j;
+        int oci = i;
+        unsigned count  = 0;
 
-        int count   = 0;
         int shift_x = 0;
         int shift_y = 0;
 
-        for(int wj=-radius; wj<=radius; ++wj) {
+#pragma unroll
+        for (int ch=0; ch<channels; ++ch)
+            means[ch] = 0;
+
+        for (int wj=-radius; wj<=radius; ++wj) {
             int hit_count = 0;
+            int tj = j + wj;
+
+            if (tj<0 || tj>dim1LenLmt) break;
 
             for(int wi=-radius; wi<=radius; ++wi) {
 
-                int tj = j + wj;
                 int ti = i + wi;
 
-                // proceed
-                float norm = 0.0f;
+                if (ti<0 || ti>dim0LenLmt) break;
+
+                AccType norm = 0;
 #pragma unroll
-                for(int ch=0; ch<channels; ++ch) {
-                    tmpclrs[ch] = shrdMem[lIdx(ti, tj, shrdLen, 1)+ch*schStride];
-                    norm += (centers[ch]-tmpclrs[ch]) * (centers[ch]-tmpclrs[ch]);
+                for (int ch=0; ch<channels; ++ch) {
+                    tmpclrs[ch] = iptr[ (ti*in.strides[0] + tj*in.strides[1] + ch*in.strides[2]) ];
+                    AccType diff = (AccType)centers[ch] - (AccType)tmpclrs[ch];
+                    norm += (diff * diff);
                 }
 
-                if (norm<= cvar) {
+                if (norm <= cvar) {
 #pragma unroll
-                    for(int ch=0; ch<channels; ++ch)
-                        means[ch] += tmpclrs[ch];
+                    for (int ch=0; ch<channels; ++ch)
+                        means[ch] += (AccType)tmpclrs[ch];
 
-                    shift_x += wi;
+                    shift_x += ti;
                     ++hit_count;
                 }
             }
-            count+= hit_count;
-            shift_y += wj*hit_count;
+            count += hit_count;
+            shift_y += tj*hit_count;
         }
 
-        if (count==0) { break; }
+        if (count==0) break;
+
+        const AccType fcount = 1/(AccType)count;
+
+        i = __float2int_rz(shift_x*fcount);
+        j = __float2int_rz(shift_y*fcount);
 
-        const float fcount = 1.f/count;
-        const int mean_x = (int)(shift_x*fcount+0.5f);
-        const int mean_y = (int)(shift_y*fcount+0.5f);
 #pragma unroll
-        for(int ch=0; ch<channels; ++ch)
-            means[ch] *= fcount;
+        for (int ch=0; ch<channels; ++ch)
+            means[ch] = __float2int_rz(means[ch]*fcount);
 
-        float norm = 0.f;
+        AccType norm = 0;
```

Comment:
AccType can only be a floating point type, right? This should be 0.0 or 0.0f like the old code.

Reply:
`AccType` can be double as well, if the input array is double.